internal/repository/attestations.go

```diff
@@ -0,0 +1,154 @@
+			return err
+		}
+		fromID = plumbing.ZeroHash.String()
+	}
+
+	ref, err := r.r.Reference(plumbing.ReferenceName(targetRef), true)
+	if err != nil {
+		return err
+	}
+	toID = ref.Hash().String()
+
+	// Does a reference authorization already exist for the parameters?
+	hasAuthorization := false
+	env, err := allAttestations.GetReferenceAuthorizationFor(r.r, targetRef, fromID, toID)
+	if err == nil {
+		hasAuthorization = true
+	} else if !errors.Is(err, attestations.ErrAuthorizationNotFound) {
+		return err
+	}
+
+	if !hasAuthorization {
+		// Create a new reference authorization and embed in env
+		statement, err := attestations.NewReferenceAuthorization(targetRef, fromID, toID)
+		if err != nil {
+			return err
+		}
+
+		env, err = dsse.CreateEnvelope(statement)
+		if err != nil {
+			return err
+		}
+	}
+
+	signer, err := signerverifier.NewSignerVerifierFromSecureSystemsLibFormat(keyBytes)
+	if err != nil {
+		return err
+	}
+
+	env, err = dsse.SignEnvelope(ctx, env, signer)
+	if err != nil {
+		return err
+	}
+
+	if err := allAttestations.SetReferenceAuthorization(r.r, env, targetRef, fromID, toID); err != nil {
+		return err
+	}
+
+	commitMessage := fmt.Sprintf("Add reference authorization for '%s' from '%s' to '%s'", targetRef, fromID, toID)
+
+	return allAttestations.Commit(r.r, commitMessage, signCommit)
+}
+
+// RemoveReferenceAuthorization removes a previously issued authorization for
+// the specified parameters. The issuer of the authorization is identified using
+// their key.
+func (r *Repository) RemoveReferenceAuthorization(keyBytes []byte, targetRef, fromID, toID string, signCommit bool) error {
+	key, err := tuf.LoadKeyFromBytes(keyBytes)
+	if err != nil {
+		return err
+	}
+	if key.KeyVal.Private == "" {
+		// The authorization removal process must be authenticated using the
+		// original signing key
+		return ErrNotSigningKey
+	}
+
+	targetRef, err = gitinterface.AbsoluteReference(r.r, targetRef)
+	if err != nil {
+		return err
+	}
+
+	allAttestations, err := attestations.LoadCurrentAttestations(r.r)
+	if err != nil {
+		return err
+	}
+
+	env, err := allAttestations.GetReferenceAuthorizationFor(r.r, targetRef, fromID, toID)
+	if err != nil {
+		if errors.Is(err, attestations.ErrAuthorizationNotFound) {
+			// No reference authorization at all
+			return nil
+		}
+		return err
+	}
+
+	newSignatures := []sslibdsse.Signature{}
+	for _, signature := range env.Signatures {
+		// This handles cases where the envelope may unintentionally have
+		// multiple signatures from the same key
+		if signature.KeyID != key.KeyID {
+			newSignatures = append(newSignatures, signature)
+		}
+	}
+	env.Signatures = newSignatures
+
+	if len(env.Signatures) == 0 {
+		if err := allAttestations.RemoveReferenceAuthorization(targetRef, fromID, toID); err != nil {
+			return err
+		}
+	} else {
+		if err := allAttestations.SetReferenceAuthorization(r.r, env, targetRef, fromID, toID); err != nil {
+			return err
+		}
+	}
```

Comment:
```suggestion
	if len(newSignatures) == 0 {
		// No more signatures, we can remove the ReferenceAuthorization.
		if err := allAttestations.RemoveReferenceAuthorization(targetRef, fromID, toID); err != nil {
			return err
		}
	} else {
		// There are still other signatures left, so update the ReferenceAuthorization to the new state.
		env.Signatures = newSignatures
		if err := allAttestations.SetReferenceAuthorization(r.r, env, targetRef, fromID, toID); err != nil {
			return err
		}
	}
```

Small nit here for readability - this confused me on first pass because I thought this was acting on the original signature state until I realized that this was the post-filtered state.

Reply:
Done.